File: threading/cpp/11_print_foobar_alternately.cpp
/*
 * QUESTION 11: Print FooBar Alternately (LeetCode 1115)
 * =====================================================
 * 
 * Problem: Two threads call foo() and bar() n times each.
 * Ensure output is "foobar" repeated n times.
 * 
 * Key Concepts: Thread alternation, semaphores, condition variables
 * 
 * Compile: g++ -std=c++20 -pthread 11_print_foobar_alternately.cpp -o foobar
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <semaphore>
#include <functional>

// SOLUTION 1: Semaphores
class FooBar_Sem {
    /*
     * EXPLANATION:
     * Two semaphores as turn signals:
     * - foo_sem: Initially 1 (foo goes first)
     * - bar_sem: Initially 0 (bar waits)
     * 
     * Sequence:
     * 1. foo acquires foo_sem, prints, releases bar_sem
     * 2. bar acquires bar_sem, prints, releases foo_sem
     */
private:
    int n_;
    std::binary_semaphore foo_sem_{1};  // foo starts
    std::binary_semaphore bar_sem_{0};  // bar waits

public:
    FooBar_Sem(int n) : n_(n) {}

    void foo(std::function<void()> printFoo) {
        for (int i = 0; i < n_; ++i) {
            foo_sem_.acquire();
            printFoo();
            bar_sem_.release();
        }
    }

    void bar(std::function<void()> printBar) {
        for (int i = 0; i < n_; ++i) {
            bar_sem_.acquire();
            printBar();
            foo_sem_.release();
        }
    }
};

// SOLUTION 2: Condition Variable
class FooBar_CV {
    /*
     * EXPLANATION:
     * Boolean flag indicates whose turn.
     * Each method waits for its turn, then flips flag.
     */
private:
    int n_;
    std::mutex mutex_;
    std::condition_variable cv_;
    bool foo_turn_ = true;

public:
    FooBar_CV(int n) : n_(n) {}

    void foo(std::function<void()> printFoo) {
        for (int i = 0; i < n_; ++i) {
            std::unique_lock<std::mutex> lock(mutex_);
            cv_.wait(lock, [this] { return foo_turn_; });
            printFoo();
            foo_turn_ = false;
            cv_.notify_one();
        }
    }

    void bar(std::function<void()> printBar) {
        for (int i = 0; i < n_; ++i) {
            std::unique_lock<std::mutex> lock(mutex_);
            cv_.wait(lock, [this] { return !foo_turn_; });
            printBar();
            foo_turn_ = true;
            cv_.notify_one();
        }
    }
};

// SOLUTION 3: Atomic with spin-wait
class FooBar_Atomic {
private:
    int n_;
    std::atomic<bool> foo_turn_{true};

public:
    FooBar_Atomic(int n) : n_(n) {}

    void foo(std::function<void()> printFoo) {
        for (int i = 0; i < n_; ++i) {
            while (!foo_turn_.load()) std::this_thread::yield();
            printFoo();
            foo_turn_.store(false);
        }
    }

    void bar(std::function<void()> printBar) {
        for (int i = 0; i < n_; ++i) {
            while (foo_turn_.load()) std::this_thread::yield();
            printBar();
            foo_turn_.store(true);
        }
    }
};

template<typename FooBarClass>
void test(const std::string& name, int n) {
    std::cout << "Testing " << name << ": ";
    
    FooBarClass fb(n);
    std::string output;
    std::mutex m;
    
    auto printFoo = [&]() { std::lock_guard<std::mutex> l(m); output += "foo"; };
    auto printBar = [&]() { std::lock_guard<std::mutex> l(m); output += "bar"; };
    
    // Start bar first to test synchronization
    std::thread t2([&]() { fb.bar(printBar); });
    std::thread t1([&]() { fb.foo(printFoo); });
    
    t1.join(); t2.join();
    
    std::string expected;
    for (int i = 0; i < n; ++i) expected += "foobar";
    
    std::cout << output;
    std::cout << (output == expected ? " ✓ PASSED" : " ✗ FAILED") << std::endl;
}

int main() {
    test<FooBar_Sem>("Semaphore", 3);
    test<FooBar_CV>("Condition Variable", 3);
    test<FooBar_Atomic>("Atomic", 3);
    return 0;
}

File: threading/python/11_print_foobar_alternately.py
"""
QUESTION 11: Print FooBar Alternately (LeetCode 1115)
=====================================================

Problem: Two threads call foo() and bar() n times each.
Ensure output is "foobar" repeated n times (alternating).

Key Concepts: Thread alternation, Semaphores, Condition variables
"""

import threading


class FooBarSemaphore:
    """
    Solution using Semaphores.

    EXPLANATION:
    Two semaphores act as turn signals:
    - foo_sem: Initially 1 (foo goes first)
    - bar_sem: Initially 0 (bar waits)

    Sequence:
    1. foo acquires foo_sem, prints, releases bar_sem
    2. bar acquires bar_sem, prints, releases foo_sem
    3. Repeat
    """

    def __init__(self, n: int):
        self.n = n
        self.foo_sem = threading.Semaphore(1)  # foo starts
        self.bar_sem = threading.Semaphore(0)  # bar waits

    def foo(self, printFoo):
        for _ in range(self.n):
            self.foo_sem.acquire()
            printFoo()
            self.bar_sem.release()

    def bar(self, printBar):
        for _ in range(self.n):
            self.bar_sem.acquire()
            printBar()
            self.foo_sem.release()


class FooBarCondition:
    """
    Solution using Condition Variable.

    EXPLANATION:
    Boolean flag indicates whose turn it is.
    Each method waits for its turn, then flips the flag.
    """

    def __init__(self, n: int):
        self.n = n
        self.foo_turn = True
        self.condition = threading.Condition()

    def foo(self, printFoo):
        for _ in range(self.n):
            with self.condition:
                while not self.foo_turn:
                    self.condition.wait()
                printFoo()
                self.foo_turn = False
                self.condition.notify()

    def bar(self, printBar):
        for _ in range(self.n):
            with self.condition:
                while self.foo_turn:
                    self.condition.wait()
                printBar()
                self.foo_turn = True
                self.condition.notify()


class FooBarEvent:
    """
    Solution using Events.

    EXPLANATION:
    Two events signal completion of each part.
    Must manually reset events for next iteration.
    """

    def __init__(self, n: int):
        self.n = n
        self.foo_done = threading.Event()
        self.bar_done = threading.Event()
        self.bar_done.set()  # foo goes first

    def foo(self, printFoo):
        for _ in range(self.n):
            self.bar_done.wait()
            self.bar_done.clear()
            printFoo()
            self.foo_done.set()

    def bar(self, printBar):
        for _ in range(self.n):
            self.foo_done.wait()
            self.foo_done.clear()
            printBar()
            self.bar_done.set()


def test(cls):
    print(f"\nTesting {cls.__name__}:")
    n = 3
    obj = cls(n)
    output = []

    def print_foo():
        output.append("foo")

    def print_bar():
        output.append("bar")

    t1 = threading.Thread(target=obj.foo, args=(print_foo,))
    t2 = threading.Thread(target=obj.bar, args=(print_bar,))

    # Start bar first to test synchronization
    t2.start()
    t1.start()

    t1.join()
    t2.join()

    result = ''.join(output)
    expected = "foobar" * n

    print(f"Output: {result}")
    assert result == expected, f"FAILED: expected {expected}"
    print("✓ PASSED")


if __name__ == "__main__":
    test(FooBarSemaphore)
    test(FooBarCondition)
    test(FooBarEvent)
